opt: add GenerateTrigramSimilarityInvertedIndexScans rule

mgartner opened the pull request:
#### sql: add optimizer_use_trigram_similarity_optimization setting

This commit adds a new session setting called
`optimizer_use_trigram_similarity_optimization`. It currently has no
effect, but in future commits it will allow control over a new
optimization for queries with trigram similarity filters.

Release note: None

#### sql: disable inverted scans when similarity optimization is enabled

The `GenerateInvertedIndexScans` rule is now disabled for similarity
filters on trigram indexes when the
`optimizer_use_text_similarity_optimization` session setting is enabled.
In a future commit this setting will enable a new rule that makes
`GenerateInvertedIndexScans` obsolete. There is no reason to trigger
both rules.

Release note: None

#### opt: add GenerateTrigramSimilarityInvertedIndexScans rule

The `GenerateTrigramSimilarityInvertedIndexScans` exploration rule has
been added, which index-accelerates trigram similarity filters. See the
comment above the newly added custom function for more details. In
future commits, this rule will be optimized further.

The tests for this rule break from convention: they are added to a new
file rather than to the `select` test file corresponding to the location
of the rule in `select.opt`. This proposed new method of organization is
motivated by the huge growth of the `select` file over the years. If
there is agreement on this, we can incrementally reorganize all existing
tests to match.

Release note (performance improvement): More efficient query plans are
now generated for queries with text similarity filters, e.g.,
`text_col % 'foobar'`. These plans are generated if the
`optimizer_use_trigram_similarity_optimization` session setting is
enabled, which it is by default (though it will be disabled by default
in previous versions).

#### opt: reduce scanned trigrams for similarity filters

The `GenerateSimilarityInvertedIndexScans` exploration rule now
generates plans that scan fewer trigrams. See the added code comments
for more details.

Release note: None

#### opt: consolidate logic for constraining inverted index prefix columns

The logic for building constraints on prefix columns of multi-column
inverted indexes has been refactored and consolidated.

Fixes #112675

Release note: None

mgartner committed:
sql: add optimizer_use_trigram_similarity_optimization setting

This commit adds a new session setting called
`optimizer_use_trigram_similarity_optimization`. It currently has no
effect, but in future commits it will allow control over a new
optimization for queries with trigram similarity filters.

Release note: None

mgartner committed:
sql: disable inverted scans when similarity optimization is enabled

The `GenerateInvertedIndexScans` rule is now disabled for similarity
filters on trigram indexes when the
`optimizer_use_text_similarity_optimization` session setting is enabled.
In a future commit this setting will enable a new rule that makes
`GenerateInvertedIndexScans` obsolete. There is no reason to trigger
both rules.

Release note: None

mgartner committed:
opt: add GenerateTrigramSimilarityInvertedIndexScans rule

The `GenerateTrigramSimilarityInvertedIndexScans` exploration rule has
been added, which index-accelerates trigram similarity filters. See the
comment above the newly added custom function for more details. In
future commits, this rule will be optimized further.

The tests for this rule break from convention: they are added to a new
file rather than to the `select` test file corresponding to the location
of the rule in `select.opt`. This proposed new method of organization is
motivated by the huge growth of the `select` file over the years. If
there is agreement on this, we can incrementally reorganize all existing
tests to match.

Release note (performance improvement): More efficient query plans are
now generated for queries with text similarity filters, e.g.,
`text_col % 'foobar'`. These plans are generated if the
`optimizer_use_trigram_similarity_optimization` session setting is
enabled, which it is by default (though it will be disabled by default
in previous versions).

mgartner committed:
opt: reduce scanned trigrams for similarity filters

The `GenerateSimilarityInvertedIndexScans` exploration rule now
generates plans that scan fewer trigrams. See the added code comments
for more details.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/121973)
<!-- Reviewable:end -->

mgartner commented:
The first commit is from #121937.

DrewKimball reviewed (commented):
Cool!

> Reviewed 10 of 10 files at r1, 2 of 2 files at r2, 15 of 15 files at r3, 11 of 11 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-:-Nv8p1mf8Yx6JD-6L7wG:b-89q77x)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`-- commits` line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv85jj6-6YPbdXQ1XJn):*
[nit] Can you specify that the rule is disabled only for trigram indexes?
___
*[`-- commits` line 35 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv88928CpRiE5RyS7BG):*
SGTM
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8kvga0HobTrAqo9rv:-Nv8kvgbEkwpHgQEPMZn:b20xqun) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L254)):*
> ```Go
> 		keyCtx.Columns.Init(cols[prefixColumnCount:])
> 
> 		var spans constraint.Spans
> ```

What's the reasoning for using regular instead of inverted spans? I'd like to avoid having to consider whether a Constraint is actually inverted because it's a trigram index, if we can.
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 292 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8jg_ZBTt3kp3TMg_e:-Nv8jg_ZBTt3kp3TMg_f:b-99d2jx) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L292)):*
> ```Go
> 	// Therefore, we only generate and check the unconsolidated constraint,
> 	// allowing the optimizer to plan multi-column inverted index scans in more
> 	// cases.
> ```

We do have a `Span.Split` method, FWIW.
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 335 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8lYy112KVB7viX06b:-Nv8lYy112KVB7viX06c:b-lsbx8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L335)):*
> ```Go
> 	d := memo.ExtractConstDatum(expr)
> 	ds, ok := d.(*tree.DString)
> 	return string(*ds), ok
> ```

Will this dereference panic if the datum isn't `DString`?
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 193 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8mG8U5qtNIj51EFSk:-Nv8mG8U5qtNIj51EFSl:b-a5d9h1) ([raw file](https://github.com/cockroachdb/cockroach/blob/713810f22e98387e46602cd9540d2b65e9f2cac3/pkg/sql/opt/invertedidx/inverted_index_expr.go#L193)):*
> ```Go
> // The returned constraint includes spans over the minimum number of trigrams in
> // a and b that must match in order to satisfy the similarity filter. This
> // optimization allows us to avoid scanning over all trigrams of the constant
> ```

[nit] `avoid scanning over all trigrams` sounds like we don't have to scan any trigrams. Maybe `some` is better?
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 430 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8oicL0LsYSwxtBh-f:-Nv8oicL0LsYSwxtBh-g:bggwprq) ([raw file](https://github.com/cockroachdb/cockroach/blob/713810f22e98387e46602cd9540d2b65e9f2cac3/pkg/sql/opt/invertedidx/inverted_index_expr.go#L430)):*
> ```Go
> 
> 		// If there are still trigrams to remove, remove trigrams as the end of
> 		// the slice.
> ```

Could this step be less effective because of the reordering above?
___
*[`pkg/sql/opt/memo/logical_props_builder.go` line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8B25J50dvo2Wqpgon:-Nv8B25J50dvo2Wqpgoo:b-oto38o) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/memo/logical_props_builder.go#L136)):*
> ```Go
> 		// rules that operate on inverted filter expressions. We should fix this
> 		// to prevent potential bugs in the future by not copying the table FDs
> 		// here and adding primary keys to the FDs of inverted filters.
> ```

Oof... do we have an issue for this?
___
*[`pkg/sql/sem/tree/datum.go` line 1727 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv870u60GfJlQV3XFXh:-Nv870u60GfJlQV3XFXi:bdhe0ig) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/sem/tree/datum.go#L1727)):*
> ```Go
> 		o = string(*t)
> 	case *DEncodedKey:
> 		// Allow comparison with DEncodedKeys. This is required for now because
> ```

[nit] This branch seems like it's expected, right?
___
*[`pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans` line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8lh_M-o8L3WWCXayh:-Nv8lh_M-o8L3WWCXayi:b85jk5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans#L44)):*
> ```
> 
> opt expect=GenerateTrigramSimilarityInvertedIndexScans
> SELECT * FROM trgm WHERE s % 'foo'
> ```

Can you add a test with `s % NULL`?
___
*[`pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans` line 117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv87gyl0oDqvk7Oxttb:-Nv87gyl0oDqvk7Oxttc:bnl114t) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans#L117)):*
> ```
>  │                   ├── [/'\x1220666f0001' - /'\x1220666f0001']
>  │                   ├── [/'\x12666f6f0001' - /'\x12666f6f0001']
>  │                   └── [/'\x126f6f200001' - /'\x126f6f200001']
> ```

Do you know why only four constraints were generated despite there being two filters?


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8kvga0HobTrAqo9rv:-Nv8xg6d3S37WTAs0fZw:b36j99c) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L254)):*
This is probably the most controversial decision in this PR and I went back-and-forth on it several times during implementation. Some reasons I ended up going with a regular constraint are:

1. The complexity of an inverted constraint isn't required.
2. We need a way to differentiate between these new scans and other inverted scans in the logical props builder so that we can avoid copying the underlying func deps of the table into the inverted scan (see the comment in the logical props builder). If we used an inverted span then we'd need some other flag in the scan private.
3. Pushing the logic in the 4th commit down into the rowenc library seemed cumbersome. It wouldn't necessarily live there forever, but for now that's probably the best place, and we'd have to plumb down some option.

The downsides of the approach in this PR are:

1. I had to write a new rule, and probably more new logic than I would have otherwise. If I used an inverted span I could have kept the original GenerateInvertedScans rule.
2. I had to make some changes in the statistics builder to get row counts estimates working correctly.
3. We don't support disjunctions of similarity filters on the same column, yet. (But this is relatively easy to add and I already have a prototype).

> I'd like to avoid having to consider whether a Constraint is actually inverted because it's a trigram index, if we can.

I'd push back on the framing of these constraints as being "actually inverted". IMO the complexity of inverted constraints has obfuscated their inner workings and made them a bit magical to us (at least more magical than constraints.

I think it's better to think of constraints as agnostic to forward or inverted indexes. They simply describe a set of spans to scan. The constraint has no notion of whether an inverted index is being scanned or not, and it doesn't need one.

I think this would have been more obvious if I was able to use `tree.DString` or `tree.Bytes` for the span keys instead of `tree.DEncodedKey` because the constraints would be printed with human-readable trigrams in them. For example, instead of `[/'\x1220666f0001' - /'\x1220666f0001']`, you'd see `[/' fo' - /' fo']`. (And in the long-term, I think this is achievable, we just need to change the histograms for trigram inverted indexes to have string upper-bounds instead of the raw encoded bytes—getting the correct row count estimates required using encoded bytes for constraints so that they matched the histogram upper-bounds.) The encoded key for `' fo'` is a prefix byte x12; then the ascii bytes for space, "f", and "o": x20, x66, x6f; then some trailing bytes x00, x01. The `[/' fo' - /' fo']` span would be encoded to exactly `[/'\x1220666f0001' - /'\x1220666f0001']` in execbuilder.

So, there's nothing special about these spans at all, at least for trigram inverted indexes. There might be some reason we need them for geospatial indexes, but I'm skeptical—after all, the scan can't be doing anything other than scanning some union of spans, AFAIK. Why do we need a special construct for it? If we were building trigram inverted indexes from scratch right now, I don't see why we'd go through the trouble of dealing with inverted spans when regular constraints should work just fine.

So I'd propose to frame it as: a scan is an "inverted scan" if it scans an inverted index, and a constraint describes the spans in that index to scan, whether it is inverted or not.

I think the questions we need to answer as a team, then, are:

1. Do we agree with this framing of constraints being inverted-agnostic? Or do we instead believe that inverted indexes demand a different type of constraint, fundamentally? This will determine our long-term direction.
2. What is the most pragmatic solution for this immediate PR? I can work on an alternative that uses inverted spans, if folks feel strongly. I'd estimate that the approach would have fewer lines of code changed—maybe a strong enough argument on its own—with the changes more intertwined with existing logic. If folks strongly disagree with (1) then it definitely makes sense to explore the alternative implementation.

Curious to hear your thoughts!


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8kvga0HobTrAqo9rv:-Nv99NeH8e3nO3R5X7XG:b-611r30) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L254)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This is probably the most controversial decision in this PR and I went back-and-forth on it several times during implementation. Some reasons I ended up going with a regular constraint are:

1. The complexity of an inverted constraint isn't required.
2. We need a way to differentiate between these new scans and other inverted scans in the logical props builder so that we can avoid copying the underlying func deps of the table into the inverted scan (see the comment in the logical props builder). If we used an inverted span then we'd need some other flag in the scan private.
3. Pushing the logic in the 4th commit down into the rowenc library seemed cumbersome. It wouldn't necessarily live there forever, but for now that's probably the best place, and we'd have to plumb down some option.

The downsides of the approach in this PR are:

1. I had to write a new rule, and probably more new logic than I would have otherwise. If I used an inverted span I could have kept the original GenerateInvertedScans rule.
2. I had to make some changes in the statistics builder to get row counts estimates working correctly.
3. We don't support disjunctions of similarity filters on the same column, yet. (But this is relatively easy to add and I already have a prototype).

> I'd like to avoid having to consider whether a Constraint is actually inverted because it's a trigram index, if we can.

I'd push back on the framing of these constraints as being "actually inverted". IMO the complexity of inverted constraints has obfuscated their inner workings and made them a bit magical to us (at least more magical than constraints.

I think it's better to think of constraints as agnostic to forward or inverted indexes. They simply describe a set of spans to scan. The constraint has no notion of whether an inverted index is being scanned or not, and it doesn't need one.

I think this would have been more obvious if I was able to use `tree.DString` or `tree.Bytes` for the span keys instead of `tree.DEncodedKey` because the constraints would be printed with human-readable trigrams in them. For example, instead of `[/'\x1220666f0001' - /'\x1220666f0001']`, you'd see `[/' fo' - /' fo']`. (And in the long-term, I think this is achievable, we just need to change the histograms for trigram inverted indexes to have string upper-bounds instead of the raw encoded bytes—getting the correct row count estimates required using encoded bytes for constraints so that they matched the histogram upper-bounds.) The encoded key for `' fo'` is a prefix byte x12; then the ascii bytes for space, "f", and "o": x20, x66, x6f; then some trailing bytes x00, x01. The `[/' fo' - /' fo']` span would be encoded to exactly `[/'\x1220666f0001' - /'\x1220666f0001']` in execbuilder.

So, there's nothing special about these spans at all, at least for trigram inverted indexes. There might be some reason we need them for geospatial indexes, but I'm skeptical—after all, the scan can't be doing anything other than scanning some union of spans, AFAIK. Why do we need a special construct for it? If we were building trigram inverted indexes from scratch right now, I don't see why we'd go through the trouble of dealing with inverted spans when regular constraints should work just fine.

So I'd propose to frame it as: a scan is an "inverted scan" if it scans an inverted index, and a constraint describes the spans in that index to scan, whether it is inverted or not.

I think the questions we need to answer as a team, then, are:

1. Do we agree with this framing of constraints being inverted-agnostic? Or do we instead believe that inverted indexes demand a different type of constraint, fundamentally? This will determine our long-term direction.
2. What is the most pragmatic solution for this immediate PR? I can work on an alternative that uses inverted spans, if folks feel strongly. I'd estimate that the approach would have fewer lines of code changed—maybe a strong enough argument on its own—with the changes more intertwined with existing logic. If folks strongly disagree with (1) then it definitely makes sense to explore the alternative implementation.

Curious to hear your thoughts!
</blockquote></details>

Just one more thought on why inverted spans might be the wrong abstraction for trigram indexes (and maybe a poor abstraction for scan spans in general):

Their complexity and versatility make it easy for poor plans to be constructed. For example:

```sql
CREATE TABLE t (
  k INT PRIMARY KEY,
  s STRING,
  INVERTED INDEX (s gin_trgm_ops)
);

EXPLAIN (OPT, VERBOSE)
SELECT * FROM t WHERE s % 'foo' AND s % 'bar';
--   select
--    ├── columns: k:1 s:2
--    ├── stable
--    ├── stats: [rows=111.1111]
--    ├── cost: 842.766667
--    ├── key: (1)
--    ├── fd: (1)-->(2)
--    ├── distribution: us-east1
--    ├── prune: (1)
--    ├── index-join t
--    │    ├── columns: k:1 s:2
--    │    ├── stats: [rows=111.1111]
--    │    ├── cost: 841.615556
--    │    ├── key: (1)
--    │    ├── fd: (1)-->(2)
--    │    ├── distribution: us-east1
--    │    └── inverted-filter
--    │         ├── columns: k:1
--    │         ├── inverted expression: /5
--    │         │    ├── tight: false, unique: false
--    │         │    ├── union spans: empty
--    │         │    └── INTERSECTION
--    │         │         ├── span expression
--    │         │         │    ├── tight: false, unique: false
--    │         │         │    └── union spans
--    │         │         │         ├── ["\x12  f\x00\x01", "\x12  f\x00\x01"]
--    │         │         │         ├── ["\x12 fo\x00\x01", "\x12 fo\x00\x01"]
--    │         │         │         ├── ["\x12foo\x00\x01", "\x12foo\x00\x01"]
--    │         │         │         └── ["\x12oo \x00\x01", "\x12oo \x00\x01"]
--    │         │         └── span expression
--    │         │              ├── tight: false, unique: false
--    │         │              └── union spans
--    │         │                   ├── ["\x12  b\x00\x01", "\x12  b\x00\x01"]
--    │         │                   ├── ["\x12 ba\x00\x01", "\x12 ba\x00\x01"]
--    │         │                   ├── ["\x12ar \x00\x01", "\x12ar \x00\x01"]
--    │         │                   └── ["\x12bar\x00\x01", "\x12bar\x00\x01"]
--    │         ├── stats: [rows=111.1111]
--    │         ├── cost: 167.151111
--    │         ├── key: (1)
--    │         ├── distribution: us-east1
--    │         └── scan t@t_s_idx
--    │              ├── columns: k:1 s_inverted_key:5
--    │              ├── inverted constraint: /5/1
--    │              │    └── spans
--    │              │         ├── ["\x12  b\x00\x01", "\x12  b\x00\x01"]
--    │              │         ├── ["\x12  f\x00\x01", "\x12  f\x00\x01"]
--    │              │         ├── ["\x12 ba\x00\x01", "\x12 ba\x00\x01"]
--    │              │         ├── ["\x12 fo\x00\x01", "\x12 fo\x00\x01"]
--    │              │         ├── ["\x12ar \x00\x01", "\x12ar \x00\x01"]
--    │              │         ├── ["\x12bar\x00\x01", "\x12bar\x00\x01"]
--    │              │         ├── ["\x12foo\x00\x01", "\x12foo\x00\x01"]
--    │              │         └── ["\x12oo \x00\x01", "\x12oo \x00\x01"]
--    │              ├── stats: [rows=111.1111, distinct(1)=111.111, null(1)=0, distinct(5)=100, null(5)=0]
--    │              ├── cost: 166.02
--    │              ├── key: (1)
--    │              ├── fd: (1)-->(5)
--    │              └── distribution: us-east1
--    └── filters
--         ├── s:2 % 'foo' [outer=(2), stable]
--         └── s:2 % 'bar' [outer=(2), stable]
```

Notice that were scanning all the trigrams of "foo" and "bar", but we only need to scan trigrams for one of those terms. The flexibility of representing conjunctions in inverted spans hurts us here. It makes me wonder what other inefficient query plans we are generating because inverted span conjunctions are naively converted to disjunctions at the scan level.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121973)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans` line 117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv87gyl0oDqvk7Oxttb:-Nv9Hj6-8XufrmgXEtqb:b-mmv934) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans#L117)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Do you know why only four constraints were generated despite there being two filters?
</blockquote></details>

Actually, I understand this now. I was thinking of it as a disjunction rather than a conjunction.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
>We need a way to differentiate between these new scans and other inverted scans in the logical props builder so that we can avoid copying the underlying func deps of the table into the inverted scan (see the comment in the logical props builder). If we used an inverted span then we'd need some other flag in the scan private.

I'd be in favor of tracking the index type explicitly in general - it seems odd and confusing that we rely on looking at the column type, anyway. It could come in handy if we ever want to add an index type that isn't distinguished by the type it operates on.

>So, there's nothing special about these spans at all, at least for trigram inverted indexes. There might be some reason we need them for geospatial indexes, but I'm skeptical—after all, the scan can't be doing anything other than scanning some union of spans, AFAIK. Why do we need a special construct for it? If we were building trigram inverted indexes from scratch right now, I don't see why we'd go through the trouble of dealing with inverted spans when regular constraints should work just fine.

>Do we agree with this framing of constraints being inverted-agnostic? Or do we instead believe that inverted indexes demand a different type of constraint, fundamentally? This will determine our long-term direction.

Maybe it comes down to the spans - while regular spans use Datums for the bounds, inverted spans are always byte slices, and can use a custom encoding (e.g. for JSON). Relevant comment [here](https://github.com/cockroachdb/cockroach/blob/de779febe1cf05b5518537716c51f34f47f42aaf/pkg/sql/inverted/expression.go#L34-L35). But then again, if we can just use that special `tree.DEncodedKey` Datum, maybe it works out anyway.

I like the idea of making inverted scans as "mundane" as possible. Though in that case, maybe we should rename inverted constraints to be "encoded" constraints, or something like that. Maybe we should prototype converting other index types over to normal constraints.

>What is the most pragmatic solution for this immediate PR? I can work on an alternative that uses inverted spans, if folks feel strongly. I'd estimate that the approach would have fewer lines of code changed—maybe a strong enough argument on its own—with the changes more intertwined with existing logic. If folks strongly disagree with (1) then it definitely makes sense to explore the alternative implementation.

Considering we want to backport this change, it would be nice to make less changes at once. But I think the optimization you showed [here](https://github.com/cockroachdb/cockroach/pull/121973#issuecomment-2048554373) will be necessary to backport as well. So the question is, is it more risky switching to normal constraints in this PR, or using inverted constraints and making some additional change to optimize the cases we care about? I suspect the former will be easier. If that's true, I think I'm in favor of keeping the current approach.

michae2 reviewed (changes requested):
This is really clever!

Is there a testcase showing the limit hint being passed down? Also, is there a testcase of the empty string % the empty string?

Also some testcases with `pg_trgm.similarity_threshold` set to 0 would be good.

> Reviewed 10 of 10 files at r1, 2 of 2 files at r2, 15 of 15 files at r3, 11 of 11 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-:-Nv9aGy49Nl9vAowfRMi:b10fckr)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 220 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9Esiw5s-x87tRgNDD:-Nv9Esiw5s-x87tRgNDE:be27so7) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L220)):*
> ```Go
> 	}
> 
> 	// First, we attempt to build a constraint from a  similarity filter on the
> ```

typo: extra space
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 242 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9J4uZ3SE1Nw15DAYn:-Nv9J4uZ3SE1Nw15DAYo:b-jvuqbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L242)):*
> ```Go
> 		s, ok := extractConstStringDatum(constStr)
> 		if !ok {
> 			continue
> ```

It would be nice if we could also handle expressions of the form `s % t` and build the spans at runtime. Maybe generic query plans will help make that possible?
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8kvga0HobTrAqo9rv:-Nv9S5WZ6AfwTmz-TRj0:b-jjin8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L254)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Just one more thought on why inverted spans might be the wrong abstraction for trigram indexes (and maybe a poor abstraction for scan spans in general):

Their complexity and versatility make it easy for poor plans to be constructed. For example:

```sql
CREATE TABLE t (
  k INT PRIMARY KEY,
  s STRING,
  INVERTED INDEX (s gin_trgm_ops)
);

EXPLAIN (OPT, VERBOSE)
SELECT * FROM t WHERE s % 'foo' AND s % 'bar';
--   select
--    ├── columns: k:1 s:2
--    ├── stable
--    ├── stats: [rows=111.1111]
--    ├── cost: 842.766667
--    ├── key: (1)
--    ├── fd: (1)-->(2)
--    ├── distribution: us-east1
--    ├── prune: (1)
--    ├── index-join t
--    │    ├── columns: k:1 s:2
--    │    ├── stats: [rows=111.1111]
--    │    ├── cost: 841.615556
--    │    ├── key: (1)
--    │    ├── fd: (1)-->(2)
--    │    ├── distribution: us-east1
--    │    └── inverted-filter
--    │         ├── columns: k:1
--    │         ├── inverted expression: /5
--    │         │    ├── tight: false, unique: false
--    │         │    ├── union spans: empty
--    │         │    └── INTERSECTION
--    │         │         ├── span expression
--    │         │         │    ├── tight: false, unique: false
--    │         │         │    └── union spans
--    │         │         │         ├── ["\x12  f\x00\x01", "\x12  f\x00\x01"]
--    │         │         │         ├── ["\x12 fo\x00\x01", "\x12 fo\x00\x01"]
--    │         │         │         ├── ["\x12foo\x00\x01", "\x12foo\x00\x01"]
--    │         │         │         └── ["\x12oo \x00\x01", "\x12oo \x00\x01"]
--    │         │         └── span expression
--    │         │              ├── tight: false, unique: false
--    │         │              └── union spans
--    │         │                   ├── ["\x12  b\x00\x01", "\x12  b\x00\x01"]
--    │         │                   ├── ["\x12 ba\x00\x01", "\x12 ba\x00\x01"]
--    │         │                   ├── ["\x12ar \x00\x01", "\x12ar \x00\x01"]
--    │         │                   └── ["\x12bar\x00\x01", "\x12bar\x00\x01"]
--    │         ├── stats: [rows=111.1111]
--    │         ├── cost: 167.151111
--    │         ├── key: (1)
--    │         ├── distribution: us-east1
--    │         └── scan t@t_s_idx
--    │              ├── columns: k:1 s_inverted_key:5
--    │              ├── inverted constraint: /5/1
--    │              │    └── spans
--    │              │         ├── ["\x12  b\x00\x01", "\x12  b\x00\x01"]
--    │              │         ├── ["\x12  f\x00\x01", "\x12  f\x00\x01"]
--    │              │         ├── ["\x12 ba\x00\x01", "\x12 ba\x00\x01"]
--    │              │         ├── ["\x12 fo\x00\x01", "\x12 fo\x00\x01"]
--    │              │         ├── ["\x12ar \x00\x01", "\x12ar \x00\x01"]
--    │              │         ├── ["\x12bar\x00\x01", "\x12bar\x00\x01"]
--    │              │         ├── ["\x12foo\x00\x01", "\x12foo\x00\x01"]
--    │              │         └── ["\x12oo \x00\x01", "\x12oo \x00\x01"]
--    │              ├── stats: [rows=111.1111, distinct(1)=111.111, null(1)=0, distinct(5)=100, null(5)=0]
--    │              ├── cost: 166.02
--    │              ├── key: (1)
--    │              ├── fd: (1)-->(5)
--    │              └── distribution: us-east1
--    └── filters
--         ├── s:2 % 'foo' [outer=(2), stable]
--         └── s:2 % 'bar' [outer=(2), stable]
```

Notice that were scanning all the trigrams of "foo" and "bar", but we only need to scan trigrams for one of those terms. The flexibility of representing conjunctions in inverted spans hurts us here. It makes me wonder what other inefficient query plans we are generating because inverted span conjunctions are naively converted to disjunctions at the scan level.
</blockquote></details>

I buy your arguments, but I defer to @DrewKimball.
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 335 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8lYy112KVB7viX06b:-Nv9RdDE4LDnISm39BQ_:b-3cv5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L335)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Will this dereference panic if the datum isn't `DString`?
</blockquote></details>

And do we need to call `tree.UnwrapDOidWrapper` before the checked cast to `DString`?
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 366 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9fBuYCPMC7Y6qtjLa:-Nv9fBuYCPMC7Y6qtjLb:bb47pis) ([raw file](https://github.com/cockroachdb/cockroach/blob/713810f22e98387e46602cd9540d2b65e9f2cac3/pkg/sql/opt/invertedidx/inverted_index_expr.go#L366)):*
> ```Go
> // similarity filter is true. The minimum number of trigrams to scan is:
> //
> // |T(b)| - (⌈pg_trgm.similarity_threshold * |T(b)|⌉ - 1)
> ```

This is really, really nice.
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 388 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9h0bA3OSh1YwOO2QZ:-Nv9h0bA3OSh1YwOO2Q_:b239b8) ([raw file](https://github.com/cockroachdb/cockroach/blob/713810f22e98387e46602cd9540d2b65e9f2cac3/pkg/sql/opt/invertedidx/inverted_index_expr.go#L388)):*
> ```Go
> 	if minMatchingTrigrams < 1 {
> 		// Ensure that minMatchingTrigrams is at least one.
> 		minMatchingTrigrams = 1
> ```

If `similarityThreshold` = 0, then I'm not sure this inverted scan works in general. Seems like we would need to return all PKs? So maybe a full scan would be better?

If `len(trgms)` = 0 (from the empty string) and `similarityThreshold` > 0, seems like we want to end up with `nil` trgms?
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 884 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9YAX1B0rdRqoV0205:-Nv9Y76A6ehVNLF-lOcR:b-4py9mk) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/memo/statistics_builder.go#L884)):*
> ```Go
> 	// there is no inverted constraint. If there is an inverted contraint, the
> 	// row count is adjusted in constrainScan.
> 	if scan.IsInvertedScan(sb.md) && scan.InvertedConstraint == nil {
> ```

Should this be above the call to `constrainScan`? Seems like we could have a 1-constraint scan for some search patterns, which I think would fall into the `constrainScan` case.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 898 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9YhRk1UowxURHp2-D:-Nv9YhRk1UowxURHp2-E:bcb6n8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/memo/statistics_builder.go#L898)):*
> ```Go
> 			// the row count of the table, but should instead reflect the number
> 			// of inverted index entries.
> 			s.RowCount = max(inputHist.ValuesCount(), 1)
> ```

Should we be calling `inputHist.Filter` using the constraints?
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr_test.go` line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9iiDg4zejtrrc6Y3i:-Nv9iiDg4zejtrrc6Y3j:bb92gmq) ([raw file](https://github.com/cockroachdb/cockroach/blob/713810f22e98387e46602cd9540d2b65e9f2cac3/pkg/sql/opt/invertedidx/inverted_index_expr_test.go#L97)):*
> ```Go
> 
> 		// We can handle thresholds outside the range of (0, 1].
> 		{"foo", 0, s("  f", " fo", "foo", "oo ")},
> ```

Would be good to also add some cases with 0 and empty strings.


<!-- Sent from Reviewable.io -->

mgartner commented:
> I'd be in favor of tracking the index type explicitly in general - it seems odd and confusing that we rely on looking at the column type, anyway. It could come in handy if we ever want to add an index type that isn't distinguished by the type it operates on.

What did you have in mind? `cat.Index` already has an `IsInverted` method, which sounds like what you are describing.

> I like the idea of making inverted scans as "mundane" as possible. Though in that case, maybe we should rename inverted constraints to be "encoded" constraints, or something like that. Maybe we should prototype converting other index types over to normal constraints.

+1 I think it may be possible to get rid of "inverted constraints" entirely. In theory we should be able to translate an `inverted.Expression` into a `constraint.Constraint` for scanning. It _should_ just be a union of all possible spans to be equivalent to the current execution behavior. We can probably reduce those spans in many cases to make plans more efficient. We only need the `inverted.Expression` in the filter step, and it might not be needed, or it might be inefficient, for many operators and index types, like trigram similarity. 

The high-level opinion of mine that has been strengthened by working on this PR is that in an ideal, longer-term, future we wouldn't have inverted constraints, expressions, or spans at all—we can use other, simpler abstractions to get the same behavior.

> But I think the optimization you showed [here](https://github.com/cockroachdb/cockroach/pull/121973#issuecomment-2048554373) will be necessary to backport as well.

This was running on master, not with the change. What optimization needs to be backported?





<!-- Sent from Reviewable.io -->

DrewKimball commented:
>This was running on master, not with the change. What optimization needs to be backported?

We need some way to efficiently handle conjunctions of similarity filters, instead of turning them into disjunctions like the plan you showed. This PR does that with normal constraints; if we used inverted constraints, I think we'd have to find another way to make that optimization happen and include it with the backport.

mgartner committed:
opt: consolidate logic for constraining inverted index prefix columns

The logic for building constraints on prefix columns of multi-column
inverted indexes has been refactored and consolidated.

Release note: None

mgartner reviewed (commented):
>We need some way to efficiently handle conjunctions of similarity filters, instead of turning them into disjunctions like the plan you showed. This PR does that with normal constraints; if we used inverted constraints, I think we'd have to find another way to make that optimization happen and include it with the backport.

:+1: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-:-NvDgeVk85ZqmT7ErZXY:b-su7drl)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`-- commits` line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv85jj6-6YPbdXQ1XJn):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Can you specify that the rule is disabled only for trigram indexes?
</blockquote></details>

Done.
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 220 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9Esiw5s-x87tRgNDD:-NvDhCKb7meUbxVv3rDK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L220)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

typo: extra space
</blockquote></details>

Done.
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 242 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9J4uZ3SE1Nw15DAYn:-NvDhEG95sGXxKny2sNx:bm2y96m) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L242)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

It would be nice if we could also handle expressions of the form `s % t` and build the spans at runtime. Maybe generic query plans will help make that possible?
</blockquote></details>

You mean like in a lookup-join expression?
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 292 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8jg_ZBTt3kp3TMg_e:-NvDSqAsDM0Wlm_t_j3h:b-cxy08s) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L292)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

We do have a `Span.Split` method, FWIW.
</blockquote></details>

This logic is actually copied from `constrainPrefixColumns`. I added a new commit which consolidates this logic. I probably won't backport that though, to avoid risk. We can consider using `Span.Split` in the future.
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 335 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8lYy112KVB7viX06b:-NvDTaTA6xvRHB_1H8OZ:bbattzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/invertedidx/inverted_index_expr.go#L335)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

And do we need to call `tree.UnwrapDOidWrapper` before the checked cast to `DString`?
</blockquote></details>

Good catches! I've fix this and added a test that would catch both: `SELECT * FROM t WHERE s % 'foo'::NAME`.
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 193 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8mG8U5qtNIj51EFSk:-NvDgtWn8MV0lWEjw0TK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/713810f22e98387e46602cd9540d2b65e9f2cac3/pkg/sql/opt/invertedidx/inverted_index_expr.go#L193)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] `avoid scanning over all trigrams` sounds like we don't have to scan any trigrams. Maybe `some` is better?
</blockquote></details>

Done.
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 366 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9fBuYCPMC7Y6qtjLa:-NvDVmBi2taxRwtTfQjx:bupmah5) ([raw file](https://github.com/cockroachdb/cockroach/blob/713810f22e98387e46602cd9540d2b65e9f2cac3/pkg/sql/opt/invertedidx/inverted_index_expr.go#L366)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

This is really, really nice.
</blockquote></details>

:tipping_hand_man:
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 388 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9h0bA3OSh1YwOO2QZ:-NvDaRKV23uqCylXPwEt:ban1vou) ([raw file](https://github.com/cockroachdb/cockroach/blob/713810f22e98387e46602cd9540d2b65e9f2cac3/pkg/sql/opt/invertedidx/inverted_index_expr.go#L388)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

If `similarityThreshold` = 0, then I'm not sure this inverted scan works in general. Seems like we would need to return all PKs? So maybe a full scan would be better?

If `len(trgms)` = 0 (from the empty string) and `similarityThreshold` > 0, seems like we want to end up with `nil` trgms?
</blockquote></details>

Great catch! I've fixed this so that we don't try to generate these plans when there are no trigrams or when the threshold is zero. I've added some tests for these cases too.
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr.go` line 430 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8oicL0LsYSwxtBh-f:-NvDaxVs4QUA1vykZnT0:bdqjav0) ([raw file](https://github.com/cockroachdb/cockroach/blob/713810f22e98387e46602cd9540d2b65e9f2cac3/pkg/sql/opt/invertedidx/inverted_index_expr.go#L430)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could this step be less effective because of the reordering above?
</blockquote></details>

It may not remove the most common trigrams, but I think it's good enough—all trigrams with spaces should have already been removed.
___
*[`pkg/sql/opt/memo/logical_props_builder.go` line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8B25J50dvo2Wqpgon:-NvDeVfH-j3Uv72Go4Zq:brqla0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/memo/logical_props_builder.go#L136)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Oof... do we have an issue for this?
</blockquote></details>

I created #122225 to track this and I've updated the TODO to point to it.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 884 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9YAX1B0rdRqoV0205:-NvDimHY51u2xyT9ZgUe:b-e1wxfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/memo/statistics_builder.go#L884)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should this be above the call to `constrainScan`? Seems like we could have a 1-constraint scan for some search patterns, which I think would fall into the `constrainScan` case.
</blockquote></details>

Amazing catch! I've fixed this and added a `memo/testdata/stats` test that catches it.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 898 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9YhRk1UowxURHp2-D:-NvDieBs9nzPtWk-hKrT:b-7xqpi9) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/memo/statistics_builder.go#L898)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should we be calling `inputHist.Filter` using the constraints?
</blockquote></details>

That happens below in the invocations of `sb.constrainScan(scan, &c, pred, relProps, &spanStats)`.
___
*[`pkg/sql/sem/tree/datum.go` line 1727 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv870u60GfJlQV3XFXh:-NvDf-vU6-oAmaCe148O:b-wwf4pv) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/sem/tree/datum.go#L1727)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] This branch seems like it's expected, right?
</blockquote></details>

Good catch. Moved it up to the previous case.
___
*[`pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans` line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8lh_M-o8L3WWCXayh:-NvDf_nB0XIqWpFfDrk0:bv2e3a1) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans#L44)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Can you add a test with `s % NULL`?
</blockquote></details>

Added. Though it's interesting, it looks like we build the canonical expression without any similarity filter. I guess it's normalized-away before normalization rules, somehow:

```
project
 ├── columns: k:1!null i:2!null j:3 s:4
 ├── immutable
 ├── key: (1)
 ├── fd: (1)-->(2-4)
 └── select
      ├── columns: k:1!null i:2!null j:3 s:4 crdb_internal_mvcc_timestamp:5 tableoid:6
      ├── immutable
      ├── key: (1)
      ├── fd: (1)-->(2-6)
      ├── scan trgm
      │    ├── columns: k:1!null i:2!null j:3 s:4 crdb_internal_mvcc_timestamp:5 tableoid:6
      │    ├── check constraint expressions
      │    │    └── i:2 IN (1, 2, 3) [outer=(2), constraints=(/2: [/1 - /1] [/2 - /2] [/3 - /3]; tight)]
      │    ├── key: (1)
      │    └── fd: (1)-->(2-6)
      └── filters
           └── NULL::BOOL [immutable]
```
___
*[`pkg/sql/opt/invertedidx/inverted_index_expr_test.go` line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv9iiDg4zejtrrc6Y3i:-NvDgYUr77vYMzIRswtt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/713810f22e98387e46602cd9540d2b65e9f2cac3/pkg/sql/opt/invertedidx/inverted_index_expr_test.go#L97)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Would be good to also add some cases with 0 and empty strings.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> This is great work!

> Reviewed 23 of 23 files at r5, 2 of 2 files at r6, 8 of 15 files at r7, 11 of 11 files at r10, 11 of 11 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-:-NvJpcDcBkHoX8rP8ptr:b-wg6qme)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans` line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8lh_M-o8L3WWCXayh:-NvJjt6m0pZNhZi9G02k:b-blp3u3) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans#L44)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Added. Though it's interesting, it looks like we build the canonical expression without any similarity filter. I guess it's normalized-away before normalization rules, somehow:

```
project
 ├── columns: k:1!null i:2!null j:3 s:4
 ├── immutable
 ├── key: (1)
 ├── fd: (1)-->(2-4)
 └── select
      ├── columns: k:1!null i:2!null j:3 s:4 crdb_internal_mvcc_timestamp:5 tableoid:6
      ├── immutable
      ├── key: (1)
      ├── fd: (1)-->(2-6)
      ├── scan trgm
      │    ├── columns: k:1!null i:2!null j:3 s:4 crdb_internal_mvcc_timestamp:5 tableoid:6
      │    ├── check constraint expressions
      │    │    └── i:2 IN (1, 2, 3) [outer=(2), constraints=(/2: [/1 - /1] [/2 - /2] [/3 - /3]; tight)]
      │    ├── key: (1)
      │    └── fd: (1)-->(2-6)
      └── filters
           └── NULL::BOOL [immutable]
```
</blockquote></details>

Maybe try with `NULL::STRING` and `disable=(FoldNullBinaryLeft,FoldNullBinaryRight)`?


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121973)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans` line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121973#-Nv8lh_M-o8L3WWCXayh:-NvXsXRe5aPUndYLaZLN:b-tb4n8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/902cdaee2c90c2951939de08489caeb053bfd185/pkg/sql/opt/xform/testdata/rules/generate_trigram_similarity_inverted_index_scans#L44)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Maybe try with `NULL::STRING` and `disable=(FoldNullBinaryLeft,FoldNullBinaryRight)`?
</blockquote></details>

Nice idea. Done.


<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+

mgartner commented:
bors r+
